### PR TITLE
Pure state sync refactoring (part-2)

### PR DESCRIPTION
**Pull request by liuchengxu**
This PR is the second part of the pure state sync refactoring, encapsulating `StateSyncMetadata` as a separate entity. Now it's pretty straightforward what changes are needed for the persistent state sync as observed in the struct `StateSync`:

- `state`: redirect directly to the DB layer instead of being accumulated in the memory.
- `metadata`: handle the state sync metadata on disk whenever the state is forwarded to the DB, resume an ongoing state sync on a restart, etc.

### CODE REVIEW

File: substrate/client/network/sync/src/strategy/state_sync.rs [hunk 1]
```diff
@@ -89,22 +89,61 @@ pub enum ImportResult<B: BlockT> {
...
+				StateSyncPhase::ImportingState
+			} else {
+				StateSyncPhase::DownloadingState
+			},
+		}
+	}
+}
+
+/// State sync state machine. Accumulates partial state data until it
+/// is ready to be imported.
```
Comment (bkchr) [inline substrate/client/network/sync/src/strategy/state_sync.rs:140]:
```suggestion
/// State sync state machine. 
///
//// Accumulates partial state data until it is ready to be imported.
```

**Review by bkchr — APPROVED** [review]

File: substrate/client/network/sync/src/strategy/state_sync.rs [hunk 2]
```diff
@@ -89,22 +89,62 @@ pub enum ImportResult<B: BlockT> {
...
+			} else {
+				StateSyncPhase::DownloadingState
+			},
+		}
+	}
+}
+
+/// State sync state machine. 
+///
+//// Accumulates partial state data until it is ready to be imported.
```
Comment (lexnv) [inline substrate/client/network/sync/src/strategy/state_sync.rs:141]:
```suggestion
/// Accumulates partial state data until it is ready to be imported.
```

File: substrate/client/network/sync/src/strategy/state_sync.rs [hunk 3]
```diff
@@ -89,22 +89,62 @@ pub enum ImportResult<B: BlockT> {
 	BadResponse,
 }
 
-/// State sync state machine. Accumulates partial state data until it
-/// is ready to be imported.
-pub struct StateSync<B: BlockT, Client> {
-	target_block: B::Hash,
+struct StateSyncMetadata<B: BlockT> {
```
Comment (lexnv) [inline substrate/client/network/sync/src/strategy/state_sync.rs:92]: minor nit; Would be useful to have some documatation on this struct and exposed methods

**Review by lexnv — APPROVED** [review]
Looks good! Thanks again for contributing 🙏

**Pull request merged**